theano/tensor/nnet/nnet.py

```diff
@@ -1784,6 +1803,59 @@ def graph_merge_softmax_with_crossentropy_softmax(node):
                     return [mergeable_client[1]]
 
 
+@opt.register_specialize
+@opt.register_stabilize
+@opt.register_canonicalize
+@gof.local_optimizer([CrossentropySoftmax1HotWithBiasDx])
+def local_useless_crossentropy_softmax_1hot_with_bias_dx_alloc(node):
+    """
+    Replaces a CrossentropySoftmax1HotWithBiasDx op, whose incoming gradient is
```

Comment:
I went through the tests, which were crashing because the optimization yields shorter graphs, and checked them manually for correctness. Then I modified the asserts of the graph length accordingly. There are still bugs in this optimization, e.g. the Assert op is actually not imported, but this case was never triggered. I've made some fixes locally, but wanted to discuss the general problem concerning the shape inference problem first.